L1Trigger/Phase2L1ParticleFlow/plugins/L1TPFProducer.cc

```diff
@@ -0,0 +1,381 @@
+    for (int i = 0; i < l1tpf_impl::Region::n_output_types; ++i) {
+      produces<unsigned int>(std::string(tot ? "totNL1PF" : "maxNL1PF") + l1tpf_impl::Region::outputTypeName(i));
+      produces<unsigned int>(std::string(tot ? "totNL1Puppi" : "maxNL1Puppi") + l1tpf_impl::Region::outputTypeName(i));
+    }
+  }
+  for (int i = 0; i < l1tpf_impl::Region::n_input_types; ++i) {
+    produces<std::vector<unsigned>>(std::string("vecNL1") + l1tpf_impl::Region::inputTypeName(i));
+  }
+  for (int i = 0; i < l1tpf_impl::Region::n_output_types; ++i) {
+    produces<std::vector<unsigned>>(std::string("vecNL1PF") + l1tpf_impl::Region::outputTypeName(i));
+    produces<std::vector<unsigned>>(std::string("vecNL1Puppi") + l1tpf_impl::Region::outputTypeName(i));
+  }
+}
+
+L1TPFProducer::~L1TPFProducer() {
+  // do anything here that needs to be done at desctruction time
+  // (e.g. close files, deallocate resources etc.)
+  if (fRegionDump)
+    fclose(fRegionDump);
+  if (fRegionCOE)
+    fRegionCOE->close();
+}
+
+// ------------ method called to produce the data  ------------
+void L1TPFProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  /// ------ READ TRACKS ----
+  if (hasTracks_) {
+    edm::Handle<l1t::PFTrackCollection> htracks;
+    iEvent.getByToken(tkCands_, htracks);
+    const auto& tracks = *htracks;
+    for (unsigned int itk = 0, ntk = tracks.size(); itk < ntk; ++itk) {
+      const auto& tk = tracks[itk];
+      // adding objects to PF
+      if (debugR_ > 0 && deltaR(tk.eta(), tk.phi(), debugEta_, debugPhi_) > debugR_)
+        continue;
+      if (tk.pt() > trkPt_ && tk.nStubs() >= trkMinStubs_ && tk.normalizedChi2() < trkMaxChi2_) {
+        l1regions_.addTrack(tk, l1t::PFTrackRef(htracks, itk));
+      }
+    }
+  }
+
+  /// ------ READ MUONS ----
+  /// ------- first check that not more than one version of muons (standaloneMu or trackerMu) is set to be used in l1pflow
+  if (useStandaloneMuons_ && useTrackerMuons_) {
+    throw cms::Exception(
+        "Configuration",
+        "setting useStandaloneMuons=True && useTrackerMuons=True is not to be done, as it would duplicate all muons\n");
+  }
+
+  if (useStandaloneMuons_) {
+    edm::Handle<l1t::MuonBxCollection> muons;
+    iEvent.getByToken(muCands_, muons);
+    for (auto it = muons->begin(0), ed = muons->end(0); it != ed; ++it) {
+      const l1t::Muon& mu = *it;
+      if (debugR_ > 0 && deltaR(mu.eta(), mu.phi(), debugEta_, debugPhi_) > debugR_)
+        continue;
+      l1regions_.addMuon(mu, l1t::PFCandidate::MuonRef(muons, muons->key(it)));
+    }
+  }
+
+  if (useTrackerMuons_) {
+    edm::Handle<l1t::TkMuonCollection> muons;
+    iEvent.getByToken(tkMuCands_, muons);
+    for (auto it = muons->begin(), ed = muons->end(); it != ed; ++it) {
+      const l1t::TkMuon& mu = *it;
+      if (debugR_ > 0 && deltaR(mu.eta(), mu.phi(), debugEta_, debugPhi_) > debugR_)
+        continue;
+      l1regions_.addMuon(mu);  // FIXME add a l1t::PFCandidate::MuonRef
```

Comment:
it would require some restructuring of the dataformats as the TkMuon is not a Muon, so while I hope it can be implemented in the future it's not in the scope of this PR  (the production instance of this uses the standalone muons and not the TkMuons)